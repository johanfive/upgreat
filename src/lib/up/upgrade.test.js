const child_process = require('../../util/child_process')
jest.mock('../../util/child_process')
const fs = require('../../util/fs')
jest.mock('../../util/fs')
const runTests = require('./runTests')
jest.mock('./runTests')
const runBuild = require('./runBuild')
jest.mock('./runBuild')

const upgrade = require('./upgrade')

const mockDep = {
  name: 'dep',
  version: '1.1.1',
  targetVersion: '2.2.2',
  diff: 'major',
}

//
;[true, false].map(isNpm => {
  describe(`${isNpm ? 'npm' : 'yarn'}`, () => {
    let upgradeFn
    const log = jest.fn()

    beforeEach(() => {
      log.mockClear()
      child_process.exec.mockClear()
      upgradeFn = upgrade(log, {
        npm: isNpm,
        testScipt: 'test',
        buildScript: 'build',
      })
    })

    it(`should return null if upgrade fails`, async () => {
      child_process.exec.mockImplementationOnce(() => Promise.reject())

      const ret = await upgradeFn(mockDep)

      expect(ret).toMatchSnapshot()
      expect(log.mock.calls).toMatchSnapshot()
    })

    it(`should runTests if not dev dep`, async () => {
      child_process.exec.mockImplementationOnce(() => Promise.resolve())
      runTests.mockImplementationOnce(() => Promise.resolve())

      const ret = await upgradeFn(mockDep)

      expect(ret).toMatchSnapshot()
      expect(log.mock.calls).toMatchSnapshot()
    })

    it(`${isNpm ? 'npm' : 'yarn'}: should runBuild if dev dep`, async () => {
      child_process.exec.mockImplementationOnce(() => Promise.resolve())
      runBuild.mockImplementationOnce(() => Promise.resolve())

      const ret = await upgradeFn({ ...mockDep, dev: true })

      expect(ret).toMatchSnapshot()
      expect(log.mock.calls).toMatchSnapshot()
    })

    it(`should write test error to file and roll back if test fails`, async () => {
      child_process.exec.mockImplementation(() => Promise.resolve())
      runBuild.mockImplementationOnce(() =>
        Promise.reject({
          stderr: 'stderr string',
        }),
      )

      const ret = await upgradeFn({ ...mockDep, dev: true })

      expect(ret).toMatchSnapshot()
      expect(log.mock.calls).toMatchSnapshot()
      expect(child_process.exec.mock.calls).toMatchSnapshot()
    })

    it(`should write test error to file and error out if rollback fails`, async () => {
      child_process.exec.mockImplementationOnce(() => Promise.resolve())
      child_process.exec.mockImplementationOnce(() =>
        Promise.reject({
          stderr: 'exec stderr',
        }),
      )
      runBuild.mockImplementationOnce(() =>
        Promise.reject({
          stderr: 'stderr string',
        }),
      )

      const ret = await upgradeFn({ ...mockDep, dev: true })

      expect(ret).toMatchSnapshot()
      expect(log.mock.calls).toMatchSnapshot()
      expect(child_process.exec.mock.calls).toMatchSnapshot()
    })
  })
})
