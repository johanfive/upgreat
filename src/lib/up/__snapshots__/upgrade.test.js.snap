// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return null if upgrade fails 1`] = `null`;

exports[`should return null if upgrade fails 2`] = `
Array [
  Array [
    "  - dep: 1.1.1 -> 2.2.2 ([31mmajor[39m)",
  ],
  Array [
    "    failed upgrade",
  ],
]
`;

exports[`should runBuild if dev dep 1`] = `
Object {
  "dev": true,
  "diff": "major",
  "name": "dep",
  "targetVersion": "2.2.2",
  "version": "1.1.1",
}
`;

exports[`should runBuild if dev dep 2`] = `
Array [
  Array [
    "  - dep: 1.1.1 -> 2.2.2 ([31mmajor[39m)",
  ],
  Array [
    "      devDependency, running build",
  ],
]
`;

exports[`should runTests if not dev dep 1`] = `
Object {
  "diff": "major",
  "name": "dep",
  "targetVersion": "2.2.2",
  "version": "1.1.1",
}
`;

exports[`should runTests if not dev dep 2`] = `
Array [
  Array [
    "  - dep: 1.1.1 -> 2.2.2 ([31mmajor[39m)",
  ],
  Array [
    "      dependency, running tests",
  ],
]
`;

exports[`should write test error to file and error out if rollback fails 1`] = `
Object {
  "dev": true,
  "diff": "major",
  "err": Object {
    "stderr": "exec stderr",
  },
  "name": "dep",
  "targetVersion": "2.2.2",
  "version": "1.1.1",
}
`;

exports[`should write test error to file and error out if rollback fails 2`] = `
Array [
  Array [
    "  - dep: 1.1.1 -> 2.2.2 ([31mmajor[39m)",
  ],
  Array [
    "      devDependency, running build",
  ],
  Array [
    "      tests/build failed after upgrade, rolling back.. (failure details in [35m.upgreat/dep@1.1.1->2.2.2.txt[39m)",
  ],
  Array [
    "      failed rolling back!",
  ],
]
`;

exports[`should write test error to file and error out if rollback fails 3`] = `
Array [
  Array [
    "yarn upgrade dep@2.2.2",
  ],
  Array [
    "yarn upgrade dep@1.1.1",
  ],
]
`;

exports[`should write test error to file and roll back if test fails 1`] = `
Object {
  "dev": true,
  "diff": "major",
  "err": Object {
    "stderr": "stderr string",
  },
  "name": "dep",
  "targetVersion": "2.2.2",
  "version": "1.1.1",
}
`;

exports[`should write test error to file and roll back if test fails 2`] = `
Array [
  Array [
    "  - dep: 1.1.1 -> 2.2.2 ([31mmajor[39m)",
  ],
  Array [
    "      devDependency, running build",
  ],
  Array [
    "      tests/build failed after upgrade, rolling back.. (failure details in [35m.upgreat/dep@1.1.1->2.2.2.txt[39m)",
  ],
]
`;

exports[`should write test error to file and roll back if test fails 3`] = `
Array [
  Array [
    "yarn upgrade dep@2.2.2",
  ],
  Array [
    "yarn upgrade dep@1.1.1",
  ],
]
`;
